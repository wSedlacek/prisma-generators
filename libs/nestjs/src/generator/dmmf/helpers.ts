// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
