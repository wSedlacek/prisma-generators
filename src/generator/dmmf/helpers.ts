// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
