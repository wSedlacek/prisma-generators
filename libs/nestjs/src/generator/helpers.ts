import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
