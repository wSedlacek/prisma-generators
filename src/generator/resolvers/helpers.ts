import { getFieldTSType, getGraphQLType } from '../helpers';
import { DmmfDocument } from '../dmmf/dmmf-document';
import { DMMF } from '../dmmf/types';

export const generateCrudResolverClassMethodDeclaration = (
  action: DMMF.Action,
  typeName: string,
  method: DMMF.SchemaField,
  argsTypeName: string | undefined,
  collectionName: string,
  dmmfDocument: DmmfDocument,
  mapping: DMMF.Mapping
) => {
  const returnTSType = getFieldTSType(
    method.outputType,
    dmmfDocument,
    false,
    mapping.model,
    typeName
  );

  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${returnTSType}>`,
    decorators: [
      {
        name: `${action.operation}`,
        arguments: [
          `() => ${getGraphQLType(
            method.outputType,
            dmmfDocument,
            mapping.model,
            typeName
          )}`,
          `{
            ${[
              `nullable: ${!method.outputType.isRequired}`,
              `description: undefined`,
            ].join(',\n')}
          }`,
        ],
      },
    ],
    parameters: [
      {
        name: 'ctx',
        // TODO: import custom `ContextType`
        type: 'any',
        decorators: [{ name: 'Context', arguments: [] }],
      },
      ...(action.kind === 'aggregate'
        ? [
            {
              name: 'info',
              type: 'GraphQLResolveInfo',
              decorators: [{ name: 'Info', arguments: [] }],
            },
          ]
        : []),
      ...(!argsTypeName
        ? []
        : [
            {
              name: 'args',
              type: argsTypeName,
              decorators: [{ name: 'Args', arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === 'aggregate'
        ? [
            `const transformFields = (fields: Record<string, any>): Record<string, any> => {
              return Object.fromEntries(
                Object.entries(fields)
                  .filter(([key, value]) => !key.startsWith("_"))
                  .map<[string, any]>(([key, value]) => {
                    if (Object.keys(value).length === 0) {
                      return [key, true];
                    }
                    return [key, transformFields(value)];
                  }),
              );
            }`,
            `return ctx.prisma.${collectionName}.${action.kind}({
              ...args,
              ...transformFields(graphqlFields(info as any)),
            });`,
          ]
        : [
            `return plainToClass(${returnTSType
              .replace(/\[|\]/g, '')
              .replace(
                '| undefined',
                ''
              )}, await ctx.prisma.${collectionName}.${action.kind}(${
              argsTypeName ? 'args' : ''
            }) as ${getGraphQLType(
              method.outputType,
              dmmfDocument,
              mapping.model,
              typeName
            )});`,
          ],
  };
};
