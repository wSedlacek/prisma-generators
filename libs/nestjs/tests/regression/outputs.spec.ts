import { promises as fs } from 'fs';

import { generateArtifactsDirPath } from '../helpers/artifacts-dir';
import { generateCodeFromSchema } from '../helpers/generate-code';
import { prisma } from '../helpers/prisma-template';
import {
  createReadGeneratedFile,
  ReadGeneratedFile,
} from '../helpers/read-file';

describe('outputs', () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath('regression-outputs');
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  it('should properly generate output type classes', async () => {
    const schema = prisma`
      model Sample {
        intIdField    Int       @id @default(autoincrement())
        stringField   String
        floatField    Float
        booleanField  Boolean
        dateField     DateTime
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const aggregateSampleTSFile = await readGeneratedFile(
      '/resolvers/outputs/AggregateSample.ts'
    );
    const avgAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/SampleAvgAggregateOutputType.ts'
    );
    const sumAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/SampleSumAggregateOutputType.ts'
    );
    const minAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/SampleMinAggregateOutputType.ts'
    );
    const maxAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/SampleMaxAggregateOutputType.ts'
    );
    const batchPayloadTSFile = await readGeneratedFile(
      '/resolvers/outputs/BatchPayload.ts'
    );
    const outputsIndexTSFile = await readGeneratedFile(
      '/resolvers/outputs/index.ts'
    );

    expect(aggregateSampleTSFile).toMatchSnapshot('AggregateSample');
    expect(avgAggregateTSFile).toMatchSnapshot('SampleAvgAggregateOutputType');
    expect(sumAggregateTSFile).toMatchSnapshot('SampleSumAggregateOutputType');
    expect(minAggregateTSFile).toMatchSnapshot('SampleMinAggregateOutputType');
    expect(maxAggregateTSFile).toMatchSnapshot('SampleMaxAggregateOutputType');
    expect(batchPayloadTSFile).toMatchSnapshot('BatchPayload');
    expect(outputsIndexTSFile).toMatchSnapshot('outputs index');
  });

  it('should properly generate aggregate classes for renamed model', async () => {
    const schema = prisma`
      /// @@NestJS.type("Example")
      model Sample {
        intIdField    Int       @id @default(autoincrement())
        stringField   String
        floatField    Float
        booleanField  Boolean
        dateField     DateTime
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const aggregateExampleTSFile = await readGeneratedFile(
      '/resolvers/outputs/AggregateExample.ts'
    );
    const avgAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleAvgAggregateOutputType.ts'
    );
    const sumAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleSumAggregateOutputType.ts'
    );
    const minAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleMinAggregateOutputType.ts'
    );
    const maxAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleMaxAggregateOutputType.ts'
    );
    const outputsIndexTSFile = await readGeneratedFile(
      '/resolvers/outputs/index.ts'
    );

    expect(aggregateExampleTSFile).toMatchSnapshot('AggregateExample');
    expect(avgAggregateTSFile).toMatchSnapshot('ExampleAvgAggregateOutputType');
    expect(sumAggregateTSFile).toMatchSnapshot('ExampleSumAggregateOutputType');
    expect(minAggregateTSFile).toMatchSnapshot('ExampleMinAggregateOutputType');
    expect(maxAggregateTSFile).toMatchSnapshot('ExampleMaxAggregateOutputType');
    expect(outputsIndexTSFile).toMatchSnapshot('outputs index');
  });

  it('should properly generate aggregate classes for model with lowercase name', async () => {
    const schema = prisma`
      model example {
        intIdField    Int       @id @default(autoincrement())
        stringField   String
        floatField    Float
        booleanField  Boolean
        dateField     DateTime
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const aggregateExampleTSFile = await readGeneratedFile(
      '/resolvers/outputs/AggregateExample.ts'
    );
    const avgAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleAvgAggregateOutputType.ts'
    );
    const sumAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleSumAggregateOutputType.ts'
    );
    const minAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleMinAggregateOutputType.ts'
    );
    const maxAggregateTSFile = await readGeneratedFile(
      '/resolvers/outputs/ExampleMaxAggregateOutputType.ts'
    );
    const outputsIndexTSFile = await readGeneratedFile(
      '/resolvers/outputs/index.ts'
    );

    expect(aggregateExampleTSFile).toMatchSnapshot('AggregateExample');
    expect(avgAggregateTSFile).toMatchSnapshot('ExampleAvgAggregateOutputType');
    expect(sumAggregateTSFile).toMatchSnapshot('ExampleSumAggregateOutputType');
    expect(minAggregateTSFile).toMatchSnapshot('ExampleMinAggregateOutputType');
    expect(maxAggregateTSFile).toMatchSnapshot('ExampleMaxAggregateOutputType');
    expect(outputsIndexTSFile).toMatchSnapshot('outputs index');
  });
});
