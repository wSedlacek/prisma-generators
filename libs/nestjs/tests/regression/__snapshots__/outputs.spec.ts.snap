// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs should properly generate aggregate classes for model with lowercase name: AggregateExample 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";
import { ExampleAvgAggregateOutputType } from \\"../outputs/ExampleAvgAggregateOutputType\\";
import { ExampleMaxAggregateOutputType } from \\"../outputs/ExampleMaxAggregateOutputType\\";
import { ExampleMinAggregateOutputType } from \\"../outputs/ExampleMinAggregateOutputType\\";
import { ExampleSumAggregateOutputType } from \\"../outputs/ExampleSumAggregateOutputType\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class AggregateExample {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  count!: number;

  @ClassTransformer__Type(() => ExampleAvgAggregateOutputType)
  @Field(() => ExampleAvgAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  avg?: ExampleAvgAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => ExampleSumAggregateOutputType)
  @Field(() => ExampleSumAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  sum?: ExampleSumAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => ExampleMinAggregateOutputType)
  @Field(() => ExampleMinAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  min?: ExampleMinAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => ExampleMaxAggregateOutputType)
  @Field(() => ExampleMaxAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  max?: ExampleMaxAggregateOutputType | null | undefined;
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: ExampleAvgAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleAvgAggregateOutputType {
  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: ExampleMaxAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleMaxAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: ExampleMinAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleMinAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: ExampleSumAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleSumAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for model with lowercase name: outputs index 1`] = `
"export { AggregateExample } from \\"./AggregateExample\\";
export { BatchPayload } from \\"./BatchPayload\\";
export { ExampleAvgAggregateOutputType } from \\"./ExampleAvgAggregateOutputType\\";
export { ExampleMaxAggregateOutputType } from \\"./ExampleMaxAggregateOutputType\\";
export { ExampleMinAggregateOutputType } from \\"./ExampleMinAggregateOutputType\\";
export { ExampleSumAggregateOutputType } from \\"./ExampleSumAggregateOutputType\\";
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: AggregateExample 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";
import { ExampleAvgAggregateOutputType } from \\"../outputs/ExampleAvgAggregateOutputType\\";
import { ExampleMaxAggregateOutputType } from \\"../outputs/ExampleMaxAggregateOutputType\\";
import { ExampleMinAggregateOutputType } from \\"../outputs/ExampleMinAggregateOutputType\\";
import { ExampleSumAggregateOutputType } from \\"../outputs/ExampleSumAggregateOutputType\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class AggregateExample {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  count!: number;

  @ClassTransformer__Type(() => ExampleAvgAggregateOutputType)
  @Field(() => ExampleAvgAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  avg?: ExampleAvgAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => ExampleSumAggregateOutputType)
  @Field(() => ExampleSumAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  sum?: ExampleSumAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => ExampleMinAggregateOutputType)
  @Field(() => ExampleMinAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  min?: ExampleMinAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => ExampleMaxAggregateOutputType)
  @Field(() => ExampleMaxAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  max?: ExampleMaxAggregateOutputType | null | undefined;
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: ExampleAvgAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleAvgAggregateOutputType {
  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: ExampleMaxAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleMaxAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: ExampleMinAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleMinAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: ExampleSumAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class ExampleSumAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate aggregate classes for renamed model: outputs index 1`] = `
"export { AggregateExample } from \\"./AggregateExample\\";
export { BatchPayload } from \\"./BatchPayload\\";
export { ExampleAvgAggregateOutputType } from \\"./ExampleAvgAggregateOutputType\\";
export { ExampleMaxAggregateOutputType } from \\"./ExampleMaxAggregateOutputType\\";
export { ExampleMinAggregateOutputType } from \\"./ExampleMinAggregateOutputType\\";
export { ExampleSumAggregateOutputType } from \\"./ExampleSumAggregateOutputType\\";
"
`;

exports[`outputs should properly generate output type classes: AggregateSample 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";
import { SampleAvgAggregateOutputType } from \\"../outputs/SampleAvgAggregateOutputType\\";
import { SampleMaxAggregateOutputType } from \\"../outputs/SampleMaxAggregateOutputType\\";
import { SampleMinAggregateOutputType } from \\"../outputs/SampleMinAggregateOutputType\\";
import { SampleSumAggregateOutputType } from \\"../outputs/SampleSumAggregateOutputType\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class AggregateSample {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  count!: number;

  @ClassTransformer__Type(() => SampleAvgAggregateOutputType)
  @Field(() => SampleAvgAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  avg?: SampleAvgAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => SampleSumAggregateOutputType)
  @Field(() => SampleSumAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  sum?: SampleSumAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => SampleMinAggregateOutputType)
  @Field(() => SampleMinAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  min?: SampleMinAggregateOutputType | null | undefined;

  @ClassTransformer__Type(() => SampleMaxAggregateOutputType)
  @Field(() => SampleMaxAggregateOutputType, {
    nullable: true,
    description: undefined
  })
  max?: SampleMaxAggregateOutputType | null | undefined;
}
"
`;

exports[`outputs should properly generate output type classes: BatchPayload 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class BatchPayload {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  count!: number;
}
"
`;

exports[`outputs should properly generate output type classes: SampleAvgAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class SampleAvgAggregateOutputType {
  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate output type classes: SampleMaxAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class SampleMaxAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate output type classes: SampleMinAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class SampleMinAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate output type classes: SampleSumAggregateOutputType 1`] = `
"import { Field, Float, ID, Int, ObjectType } from \\"@nestjs/graphql\\";
import GraphQLJSON from \\"graphql-type-json\\";
import { JsonValue, InputJsonValue } from \\"../../../../helpers/prisma-client-mock\\";
import { Type as ClassTransformer__Type } from \\"class-transformer\\";

@ObjectType({
  isAbstract: true,
  description: undefined,
})
export class SampleSumAggregateOutputType {
  @Field(() => Int, {
    nullable: false,
    description: undefined
  })
  intIdField!: number;

  @Field(() => Float, {
    nullable: false,
    description: undefined
  })
  floatField!: number;
}
"
`;

exports[`outputs should properly generate output type classes: outputs index 1`] = `
"export { AggregateSample } from \\"./AggregateSample\\";
export { BatchPayload } from \\"./BatchPayload\\";
export { SampleAvgAggregateOutputType } from \\"./SampleAvgAggregateOutputType\\";
export { SampleMaxAggregateOutputType } from \\"./SampleMaxAggregateOutputType\\";
export { SampleMinAggregateOutputType } from \\"./SampleMinAggregateOutputType\\";
export { SampleSumAggregateOutputType } from \\"./SampleSumAggregateOutputType\\";
"
`;
